### ISSUE TITLE
AssetDataRegistry->enqueue_asset_data and  takes nearly 60% of a WooCommerce > Orders page renders

### ISSUE THREAD

**Issue by hjoelr** [woocommerce/woocommerce-admin#5690]
**Describe the bug**
Loading the WooCommerce Orders page with WC Admin enabled takes over twice as long as a page load without WC Admin enabled. This may be true of other pages as well, but I was mainly focusing on the `shop_order` page load. Here is a screenshot of an XDebug profile from a page load with WC Admin enabled.

![image](https://user-images.githubusercontent.com/221971/99683843-c6705b80-2a4e-11eb-92d1-5c82b6411138.png)

I have a few thoughts about this.

1. Is it really necessary to initialize all this data and store it as a JSON object in the browser? I understand it would remove the need for a portion of other REST API calls for the data later. However, with stores that still mainly use the Classic editor--which operates through page loads--it would be way better for any WC Admin data to be retrieved through AJAX (ie. off main thread) calls.

2. Is all the data that's being loaded really necessary? Maybe there could be a hybrid of pre-loading just enough to be useful without significantly impacting page load and then later use REST API to load anything else that's needed?

3. Can those REST API calls only load what is needed instead of initializing the entire REST API stack just to retrieve one piece of data?

4. Just initializing the REST API the first time on a call takes 23% of page load time. Automattic\WooCommerce\Blocks\Assets::get_wc_block_data() happens to be the first thing to initialize the REST API. Here's a screenshot of that call tree.

![image](https://user-images.githubusercontent.com/221971/99687862-223ce380-2a53-11eb-9f78-3061e1000ef9.png)

**To Reproduce**

1. Disable all plugins except WooCommerce 4.7.0.
2. Enable XDebug profiling
3. Load WooCommerce > Orders page
4. Disable XDebug profiling
5. Inspect XDebug profile with a tool like PhpStorm

To compare a page load without WC Admin, I added the following line to a custom plugin I had enabled just for this:

`add_filter( 'woocommerce_admin_disabled', '__return_true' );`

**Expected behavior**
I expected page load to take slightly longer (maybe 5-10%), but not over 100% slower!

**Desktop (please complete the following information):**

* OS: Windows 10 20H2
* Browser: Chrome
* Version: 86.0.4240.198

**XDebug Profile files**

- [XDebug Profile files (cachegrind.zip)](https://github.com/woocommerce/woocommerce-admin/files/5568192/cachegrind.zip)

**Comment by hjoelr** [comment]
I just realized I hadn't looked at the time it took for the page loads. You'll notice from this comparison of the two page loads (top with WC Admin enabled and bottom without WC Admin enabled) that it took in my environment 3 additional seconds for the page to load. That was around 2.7 times longer than the load without WC Admin.

![image](https://user-images.githubusercontent.com/221971/99689784-67faab80-2a55-11eb-8b11-1ceabe7a9b75.png)

**Comment by timmyc** [comment]
@hjoelr really appreciate you taking the time to do this analysis and for writing such a detailed issue, thanks! I did a quick look to see where this code might exist and I _think_ it is coming from [this portion](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/3c55aac1b599a5e4a8d2bf8ec56e89f6c56b3ad5/src/Assets/AssetDataRegistry.php#L77) of the WooCommerce Blocks code base. Going to cc @nerrad to confirm though and get any further thoughts.

**Comment by nerrad** [comment]
Took a _brief look_ and it looks like the WC_Admin is hooking into the [`woocommerce_shared_settings` filter](https://github.com/woocommerce/woocommerce-admin/blob/8862f1e88242cfa8dfb24293afa6515232ef5f32/src/Features/Onboarding.php#L222) which is exposed by the Blocks`AssetDataRegistry` for easier transfer of data between server and client. So the ultimate source of this additional load is whatever is being registered via that filter (thus triggering additional core_data from `AssetDataRegistry` to be loaded). 

There may be some overlap with the recent issue created in #5702 where the Onboarding things are loading in more places than necessary?

With that said, there may be some optimization that can be done for the Asset registry API in blocks impacting things here as well. I suspect the majority of things are coming from implementation of the [`woocommerce_shared_settings` hook](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/99cf558a9eb737676461d4f474883f91c310ca98/src/Assets/AssetDataRegistry.php#L121-L133) which was only ever intended as a back-compat shim (take note of the comment block there). Ideally we should move all usage of that filter to implement the interface in `AssetDataRegistry` for adding data as needed instead of via the filter (blocks is hooking into the filter too which is possibly what's surfacing some of the things from Blocks on these routes).

To sum up, this is both a WC Admin and Blocks issue needing addressed.

**Comment by hjoelr** [comment]
@timmyc You're welcome. Helping you guys ultimately will help me and many others.

**Comment by joshuatf** [comment]
@nerrad Is the issue a matter of switching to a different filter/method to enqueue that data or is it the fact that there is just so much preloaded data for all of these pages?

@psealock and I had a conversation earlier today about beginning the process of removing preloaded data in favor of using REST endpoints to fetch the data after load.

Ultimately we need to move away from dumping everything into preloaded data and come up with better loading interactions as well as developing thoughtful REST endpoints for data that don't yet exist.

**Comment by nerrad** [comment]
> is it the fact that there is just so much preloaded data for all of these pages?

It's mostly that there's pre-loaded data/settings loading on routes that are not using it.

- `AssetDataRegistry` makes the flawed assumption that _anytime_ this interface is used for exposing data on a route [a base set of core data is desired](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/fabf8ac5e582c5df7d8367099405e450ee9be593/src/Assets/AssetDataRegistry.php#L70). While that may be true for some of it (`wpVersion`, `wcVersion` etc), there's other stuff that probably isn't needed all the time (particularly `countries`).
- improper use of the [`woocommerce_shared_settings`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/fabf8ac5e582c5df7d8367099405e450ee9be593/src/Assets/AssetDataRegistry.php#L130-L133) filter there (which was only intended to be temporary), can result in unnecessary data loading on certain routes. This would happen when there's code that hooks into that filter on _every route_ instead of just routes needing the data provided via the callback on the filter. Any code using this filter should be switched to use the [`AssetDataRegistry::add`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/fabf8ac5e582c5df7d8367099405e450ee9be593/src/Assets/AssetDataRegistry.php#L174-L197) method (which also allows for registering a callback to call just in time for calculating/getting data on demand).

> @psealock and I had a conversation earlier today about beginning the process of removing preloaded data in favor of using REST endpoints to fetch the data after load

Ya this is definitely preferable for any data that isn't needed for the initial view (which is primarily where pre-loaded data is useful).

> Ultimately we need to move away from dumping everything into preloaded data and come up with better loading interactions as well as developing thoughtful REST endpoints for data that don't yet exist.

I agree about moving away from having _everything_ brought in via pre-loaded data, but as mentioned earlier its still useful to improve the time to first paint from the _user perspective_ (pre-hydrating etc). But again, only what's needed for the first view.

💯 on develping thoughtful REST endpoints for getting data that isn't available via REST yet because that also has the side-effect of making it easier to do headless Woo (and for mobile apps).

**Comment by ilyasfoo** [comment]
>I agree about moving away from having everything brought in via pre-loaded data, but as mentioned earlier its still useful to improve the time to first paint from the user perspective (pre-hydrating etc). But again, only what's needed for the first view.

I feel like this is a tough nut to crack since we can't know for sure what set of data the client needs from the backend's perspective. Maybe the best we could do is to have different sets of pre-loaded data for different sections of the app.

I may close this issue in a few days since it's not used to track any specific fixes.

**Comment by ilyasfoo** [comment]
Closing this issue for now
